webhook/psbinding/reconciler.go

```diff
@@ -0,0 +1,376 @@
+			"finalizers":      fb.GetFinalizers()[1:],
+			"resourceVersion": fb.GetResourceVersion(),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// ... and apply it.
+	_, err = r.DynamicClient.Resource(r.GVR).Namespace(fb.GetNamespace()).Patch(fb.GetName(),
+		types.MergePatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+// ReconcileSubject handles applying the provided Binding "mutation" (Do or
+// Undo) to the Binding's subject(s).
+func (r *BaseReconciler) ReconcileSubject(ctx context.Context, fb Bindable, mutation Mutation) error {
+
+	// Access the subject of our Binding and have the tracker queue this
+	// Bindable whenever it changes.
+	subject := fb.GetSubject()
+	if err := r.Tracker.TrackReference(subject, fb); err != nil {
+		logging.FromContext(ctx).Errorf("Error tracking subject %v: %v", subject, err)
+		return err
+	}
+
+	// Determine the GroupVersionResource of the subject reference
+	gv, err := schema.ParseGroupVersion(subject.APIVersion)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Error parsing GroupVersion %v: %v", subject.APIVersion, err)
+		return err
+	}
+	gvr := apis.KindToResource(gv.WithKind(subject.Kind))
+
+	// Use the GVR of the subject(s) to get ahold of a lister that we can
+	// use to fetch our PodSpecable resources.
+	_, lister, err := r.Factory.Get(gvr)
+	if err != nil {
+		return fmt.Errorf("error getting a lister for resource '%+v': %v", gvr, err)
+	}
+
+	// Based on the type of subject reference, build up a list of referents.
+	var referents []*duckv1.WithPod
+	if subject.Name != "" {
+		// If name is specified, then fetch it from the lister and turn
+		// it into a singleton list.
+		psObj, err := lister.ByNamespace(subject.Namespace).Get(subject.Name)
+		if apierrs.IsNotFound(err) {
+			fb.GetBindingStatus().MarkBindingUnavailable("SubjectMissing", err.Error())
+			return err
+		} else if err != nil {
+			return fmt.Errorf("error fetching Pod Speccable %v: %v", subject, err)
+		}
+		referents = append(referents, psObj.(*duckv1.WithPod))
+	} else {
+		// Otherwise, the subject is referenced by selector, so compile
+		// the selector and pass it to the lister.
+		selector, err := metav1.LabelSelectorAsSelector(subject.Selector)
+		if err != nil {
+			return err
+		}
+		psObjs, err := lister.ByNamespace(subject.Namespace).List(selector)
+		if err != nil {
+			return fmt.Errorf("error fetching Pod Speccable %v: %v", subject, err)
+		}
+		// Type cast the returned resources into our referent list.
+		for _, psObj := range psObjs {
+			referents = append(referents, psObj.(*duckv1.WithPod))
+		}
+	}
+
+	// Callback into the user's code to setup the context with additional
+	// information needed to perform the mutation.
+	if r.WithContext != nil {
+		ctx, err = r.WithContext(ctx, fb)
+		if err != nil {
+			return err
+		}
+	}
+
+	// For each of the referents, apply the mutation.
+	eg := errgroup.Group{}
+	for _, ps := range referents {
+		ps := ps
+		eg.Go(func() error {
```

Comment:
can this race? e.g. another reconcile loop operating on the same objects?

Reply:
In isolation, yes, but with the webhook it shouldn't because when the other reconcile loop attempts to undo things, it'd get filled in just like server-side defaulting and result in a nop change in etcd.